tests/catch/unit/memory/MemUtils.hh

```diff
@@ -0,0 +1,638 @@
+
+      hipExtent extent;
+      extent.width = dataW;
+      extent.height = dataH;
+      extent.depth = dataD;
+
+      params.extent = extent;
+
+      HIP_CHECK(hipMemcpy3D(&params));
+      break;
+    }
+
+    case memType::hipMem2D:
+      HIP_CHECK(hipMemcpy2D(hostMem, dataW * elementSize, aPtr, dataPitch, dataW, dataH,
+                            hipMemcpyHostToHost));
+      break;
+
+    default:
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyHostToHost));
+      break;
+  }
+}
+
+template <typename T>
+static inline void devRegisteredCopy(memType memType, T* aPtr, T* hostMem, size_t dataW,
+                                     size_t dataH, size_t dataD, size_t& dataPitch) {
+  size_t elementSize = sizeof(T);
+
+  switch (memType) {
+    case memType::hipMem3D: {
+      hipMemcpy3DParms params{};
+      params.kind = hipMemcpyHostToHost;
+      params.srcPos = make_hipPos(0, 0, 0);
+      params.dstPos = make_hipPos(0, 0, 0);
+      params.srcPtr = make_hipPitchedPtr(aPtr, dataPitch, dataW, dataH);
+      params.dstPtr = make_hipPitchedPtr(hostMem, dataW, dataW, dataH);
+
+      hipExtent extent;
+      extent.width = dataW;
+      extent.height = dataH;
+      extent.depth = dataD;
+
+      params.extent = extent;
+
+      HIP_CHECK(hipMemcpy3D(&params));
+      break;
+    }
+
+    case memType::hipMem2D:
+      HIP_CHECK(hipMemcpy2D(hostMem, dataW * elementSize, aPtr, dataPitch, dataW, dataH,
+                            hipMemcpyDeviceToHost));
+      break;
+
+    default: {
+      size_t sizeInBytes = elementSize * dataW * dataH * dataD;
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyDeviceToHost));
+      break;
+    }
+  }
+}
+
+/*
+ * function returns an offset location in memory based on the provided data, taking pitch into
+ * account
+ * (for 1D requires data.depth & data.height = 0, for 2D data.depth = 0)
+ */
+static inline size_t getPtrOffset(MultiDData data) {
+  if (data.height == 0) {  // 1D
+    return data.offset;
+  } else if (data.depth == 0) {
+    return (data.offset + (data.pitch * data.offset));
+  } else {  // 2D or 3D
+    return (data.offset + (data.pitch * data.offset) + (data.pitch * data.offset * data.height));
+  }
+}
+
+/*
+ * Description here
+ */
+
+template <typename T>
+static inline void handleHostMemcpyTypes(allocType type, memType memType, T* aPtr, MultiDData& data,
+                                         T* fillerData, size_t sizeInBytes, bool async,
+                                         hipStream_t stream) {
+  hipMemcpyKind cpyType{};
+  switch (type) {
+    case allocType::deviceMalloc:
+      cpyType = hipMemcpyHostToDevice;
+      break;
+    case allocType::devRegistered:
+      cpyType = hipMemcpyHostToDevice;
+      break;
+    default:  // host
+      cpyType = hipMemcpyHostToHost;
+      break;
+  }
+
+
+  size_t dataW = data.width;
+  size_t dataH = data.height == 0 ? 1 : data.height;
```

Comment:
What about having getters to avoid this spamming?

Reply:
Getters added to struct.